src/models/priority_queue.ts

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+/* eslint class-methods-use-this: ["error", { "exceptMethods": ["parent", "left", "right", "comparator"] }] */
+
+interface PriorityQueueNode<T> {
+  value: T;
+  priority: number;
+}
+
+/**
+ * Priority queue implementation using an underlying heap for performance.
+ * @example
+ * const prioQ = new PriorityQueue<string>();
+ * prioQ.push(prioQ.makeNode(0, 'lower prio'));
+ * prioQ.push(prioQ.makeNode(10, 'higher prio'));
+ * prioQ.push(prioQ.makeNode(5, 'medium prio'));
+ * // Returns 'higher prio'
+ * prioQ.pop()
+ * // Returns 'medium prio'
+ * prioQ.pop()
+ * // Returns 'lower prio'
+ * prioQ.pop()
+ *
+ * @template [T=unknown]
+ */
+export default class PriorityQueue<T = unknown> {
+  #heap: PriorityQueueNode<T>[];
+
+  readonly #top: number;
+
+  constructor() {
+    this.#heap = [];
+    this.#top = 0;
+  }
+
+  /**
+   * Utility to create a node from a value and priority.
+   * @example
+   * // returns { value: 'foobar', 'priority': 10 }
+   * prioQueue.makeNode(10, 'foobar');
+   */
+  static makeNode<N>(priority: number, value: N): PriorityQueueNode<N> {
+    return { value, priority };
+  }
+
+  /**
+   * Get the number of elements on the priority queue.
+   */
+  public get size() {
+    return this.#heap.length;
+  }
+
+  /**
+   * Check if the priority queue is empty
+   */
+  public isEmpty() {
+    return this.size === 0;
+  }
+
+  /**
+   * Get the top value of the queue without removing it from the queue.
+   */
+  public peek() {
+    if (this.size === 0) {
+      // Heap is empty
+      return undefined;
+    }
+    return this.#heap[this.#top].value;
+  }
+
+  /**
+   * Add a node to the priority queue and maintain the priority order.
+   */
+  public push(value: PriorityQueueNode<T>) {
+    this.#heap.push(value);
+    this._siftUp();
+  }
+
+  /**
+   * Add multiple values to the priority queue while maintaining the priority order.
+   */
+  public add(...nodes: PriorityQueueNode<T>[]) {
+    nodes.forEach(node => {
+      this.push(node);
+    });
```

Comment:
Not really, `_siftUp` will only affect the branch from the last node to the root and does not touch the other branches.

Imagine a tree of height 3, where the siftUp will work his way up from the right-most (root -> right -> right) node to the root and that there is a higher priority node on the left-most (root -> left -> left) node.
Since the `_siftUp` will see only the node and its parents up to the root, it means that when the call is over there will be a higher priority node on the bottom of the heap.

The only way to truly have a `O(logn)` insert is to insert when the priority queue is ready, meaning that all children are lower priority than the parent, which means we should be sifting up on all inserts.